### PR TITLE
🆙 Update matmul 4bit for ort>=1.17.0

### PR DESCRIPTION
**Pull request by trajepl**
## Describe your changes
Update arguments of matmul 4bit for ort>=1.17.0
1. accuracy_level
2. algorithm

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: olive/passes/onnx/quantization.py [hunk 1]
```diff
@@ -594,6 +594,29 @@ def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassCon
                 default_value=None,
                 description="List of node names to exclude from quantization.",
             ),
+            "accuracy_level": PassConfigParam(
+                type_=Any,
```
Comment (xiaoyu-work) [inline olive/passes/onnx/quantization.py:598]: Should this be `int`?

File: olive/passes/onnx/quantization.py [hunk 2]
```diff
@@ -594,6 +594,29 @@ def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassCon
...
+                description="Available from onnxruntime>=1.17.0 "
+                "The minimum accuracy level of input A, can be: 0(unset), 1(fp32), 2(fp16), 3(bf16), "
+                "or 4(int8) (default unset when 0 or None). It is used to control how input A is quantized or downcast "
+                "internally while doing computation, for example: 0 means input A will not be quantized "
+                "or downcast while doing computation. 4 means input A can be quantized with the same "
+                "block_size to int8 internally from type T1. "
+                "Refer to the MatMulNBits contrib op's 'accuracy_level' attribute for details "
+                "(https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#commicrosoftmatmulnbits).",
+            ),
+            "algorithm": PassConfigParam(
```
Comment (xiaoyu-work) [inline olive/passes/onnx/quantization.py:609]: What if the input str value is not desired? Should we either add a value validation here, or having an enum class defined for validation for this value?

**Comment by guotuofeng** [comment]
Any user requires this feature?

File: olive/passes/onnx/quantization.py (hunk 2 above)
Comment (trajepl) [reply olive/passes/onnx/quantization.py:611]: Updated.

File: olive/passes/onnx/quantization.py [hunk 3]
```diff
@@ -610,13 +642,57 @@ def _run_for_config(
...
+    if not v:
+        return v
+
+    if v not in (0, 1, 2, 3, 4):
+        raise ValueError(f"OnnxMatMul4Quantizer {field.name} must be 0(unset), 1(fp32), 2(fp16), 3(bf16) or 4(int8)")
+
+    return v
+
+
+def _validate_algorithm(v, values, field):
```
Comment (guotuofeng) [inline olive/passes/onnx/quantization.py:683]: the validate_algorithm seems has additional logic to verify neural compressor installation. Better to make sure only algo valiation logic in this function and move the validation inc logic somewhere else.

File: olive/passes/onnx/quantization.py [hunk 4]
```diff
@@ -610,13 +642,58 @@ def _run_for_config(
 
         output_model_path = resolve_onnx_path(output_model_path, Path(model.model_path).name)
 
-        quant = MatMul4BitsQuantizer(
-            model.load_model(), config["block_size"], config["is_symmetric"], config["nodes_to_exclude"]
-        )
+        if version.parse(OrtVersion) >= version.parse("1.17.0"):
+            from onnxruntime.quantization.matmul_4bits_quantizer import WeightOnlyQuantConfig
+
+            if config["algorithm"] in ("RTN", "GPTQ"):
```
Comment (guotuofeng) [inline olive/passes/onnx/quantization.py:648]: I suppose the underlying ORT will do raise the ImportError and we might need not this check?

File: olive/passes/onnx/quantization.py (hunk 3 above)
Comment (jambayk) [reply olive/passes/onnx/quantization.py:692]: saw the changes have already been made but just wanted to agree to Mikes point. Version checks and import checkers are runtime decisions to be run on the pass host. doing it during config validation or search point validation runs it on the engine's machine which is not necessarily the pass host.

File: olive/passes/onnx/quantization.py (hunk 4 above)
Comment (trajepl) [reply olive/passes/onnx/quantization.py:648]: Yes, the import error will be raised in ort as well. But no extra info to tell which config needs inc package. 
Maybe we can keep it here to provide clear description once user choose the wrong config.

**Pull request merged**